File: snippets/screens/flutter/user_profile_screen/3/user_profile_screen.dart
import 'package:flutter/material.dart';

class UserProfileScreen extends StatefulWidget {
  final String username;

  const UserProfileScreen({Key? key, required this.username}) : super(key: key);

  @override
  _UserProfileScreenState createState() => _UserProfileScreenState();
}

class _UserProfileScreenState extends State<UserProfileScreen>
    with SingleTickerProviderStateMixin {
  late TabController _tabController;
  bool isFollowing = false;
  bool isCurrentUser = false;

  final UserProfile userProfile = UserProfile(
    username: 'john_doe',
    displayName: 'John Doe',
    bio: 'Photographer 📸 | Travel enthusiast ✈️ | Coffee lover ☕️\nCapturing moments around the world',
    avatar: 'https://example.com/avatar1.jpg',
    postsCount: 127,
    followersCount: 1524,
    followingCount: 892,
    isVerified: true,
    website: 'www.johndoe.photography',
  );

  final List<Post> userPosts = [
    Post(
      id: '1',
      imageUrl: 'https://example.com/post1.jpg',
      caption: 'Beautiful sunset',
      likes: 156,
      comments: 23,
    ),
    Post(
      id: '2',
      imageUrl: 'https://example.com/post2.jpg',
      caption: 'Morning coffee',
      likes: 89,
      comments: 12,
    ),
    Post(
      id: '3',
      imageUrl: 'https://example.com/post3.jpg',
      caption: 'City lights',
      likes: 234,
      comments: 45,
    ),
    Post(
      id: '4',
      imageUrl: 'https://example.com/post4.jpg',
      caption: 'Mountain view',
      likes: 312,
      comments: 67,
    ),
    Post(
      id: '5',
      imageUrl: 'https://example.com/post5.jpg',
      caption: 'Beach day',
      likes: 198,
      comments: 34,
    ),
    Post(
      id: '6',
      imageUrl: 'https://example.com/post6.jpg',
      caption: 'Street art',
      likes: 156,
      comments: 28,
    ),
  ];

  @override
  void initState() {
    super.initState();
    _tabController = TabController(length: 3, vsync: this);
    isCurrentUser = widget.username == 'current_user';
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Row(
          children: [
            Text(
              userProfile.username,
              style: TextStyle(fontWeight: FontWeight.bold),
            ),
            if (userProfile.isVerified)
              Padding(
                padding: EdgeInsets.only(left: 4),
                child: Icon(Icons.verified, color: Colors.blue, size: 20),
              ),
          ],
        ),
        backgroundColor: Colors.white,
        foregroundColor: Colors.black,
        elevation: 1,
        actions: [
          if (isCurrentUser) ...[
            IconButton(
              icon: Icon(Icons.menu),
              onPressed: _showOptionsMenu,
            ),
          ] else ...[
            IconButton(
              icon: Icon(Icons.more_vert),
              onPressed: _showUserOptionsMenu,
            ),
          ],
        ],
      ),
      body: NestedScrollView(
        headerSliverBuilder: (context, innerBoxIsScrolled) {
          return [
            SliverToBoxAdapter(
              child: Column(
                children: [
                  _buildProfileHeader(),
                  _buildActionButtons(),
                  _buildBio(),
                  _buildHighlights(),
                  _buildTabBar(),
                ],
              ),
            ),
          ];
        },
        body: TabBarView(
          controller: _tabController,
          children: [
            _buildPostsGrid(),
            _buildReelsGrid(),
            _buildTaggedGrid(),
          ],
        ),
      ),
    );
  }

  Widget _buildProfileHeader() {
    return Padding(
      padding: EdgeInsets.all(16),
      child: Row(
        children: [
          // Profile picture
          CircleAvatar(
            radius: 40,
            backgroundImage: NetworkImage(userProfile.avatar),
          ),
          SizedBox(width: 20),
          // Stats
          Expanded(
            child: Row(
              mainAxisAlignment: MainAxisAlignment.spaceEvenly,
              children: [
                _buildStatColumn('Posts', userProfile.postsCount),
                _buildStatColumn('Followers', userProfile.followersCount),
                _buildStatColumn('Following', userProfile.followingCount),
              ],
            ),
          ),
        ],
      ),
    );
  }

  Widget _buildStatColumn(String label, int count) {
    return GestureDetector(
      onTap: () {
        if (label == 'Followers') {
          Navigator.pushNamed(context, '/followers', arguments: widget.username);
        } else if (label == 'Following') {
          Navigator.pushNamed(context, '/following', arguments: widget.username);
        }
      },
      child: Column(
        children: [
          Text(
            _formatCount(count),
            style: TextStyle(
              fontSize: 18,
              fontWeight: FontWeight.bold,
            ),
          ),
          Text(
            label,
            style: TextStyle(
              fontSize: 14,
              color: Colors.grey[600],
            ),
          ),
        ],
      ),
    );
  }

  Widget _buildActionButtons() {
    return Padding(
      padding: EdgeInsets.symmetric(horizontal: 16, vertical: 8),
      child: Row(
        children: [
          if (isCurrentUser) ...[
            Expanded(
              child: OutlinedButton(
                onPressed: _editProfile,
                child: Text('Edit Profile'),
              ),
            ),
            SizedBox(width: 8),
            Expanded(
              child: OutlinedButton(
                onPressed: _shareProfile,
                child: Text('Share Profile'),
              ),
            ),
          ] else ...[
            Expanded(
              child: ElevatedButton(
                onPressed: _toggleFollow,
                style: ElevatedButton.styleFrom(
                  backgroundColor: isFollowing ? Colors.grey[300] : Theme.of(context).primaryColor,
                  foregroundColor: isFollowing ? Colors.black : Colors.white,
                ),
                child: Text(isFollowing ? 'Following' : 'Follow'),
              ),
            ),
            SizedBox(width: 8),
            Expanded(
              child: OutlinedButton(
                onPressed: _sendMessage,
                child: Text('Message'),
              ),
            ),
          ],
        ],
      ),
    );
  }

  Widget _buildBio() {
    return Padding(
      padding: EdgeInsets.symmetric(horizontal: 16, vertical: 8),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Text(
            userProfile.displayName,
            style: TextStyle(fontWeight: FontWeight.bold),
          ),
          SizedBox(height: 4),
          Text(userProfile.bio),
          if (userProfile.website.isNotEmpty) ...[
            SizedBox(height: 4),
            GestureDetector(
              onTap: _openWebsite,
              child: Text(
                userProfile.website,
                style: TextStyle(
                  color: Colors.blue,
                  decoration: TextDecoration.underline,
                ),
              ),
            ),
          ],
        ],
      ),
    );
  }

  Widget _buildHighlights() {
    final highlights = [
      HighlightStory(id: '1', title: 'Travel', cover: 'https://example.com/highlight1.jpg'),
      HighlightStory(id: '2', title: 'Food', cover: 'https://example.com/highlight2.jpg'),
      HighlightStory(id: '3', title: 'Work', cover: 'https://example.com/highlight3.jpg'),
    ];

    return Container(
      height: 100,
      padding: EdgeInsets.symmetric(horizontal: 16, vertical: 8),
      child: ListView.builder(
        scrollDirection: Axis.horizontal,
        itemCount: highlights.length + (isCurrentUser ? 1 : 0),
        itemBuilder: (context, index) {
          if (isCurrentUser && index == 0) {
            return _buildAddHighlight();
          }
          final highlightIndex = isCurrentUser ? index - 1 : index;
          return _buildHighlightItem(highlights[highlightIndex]);
        },
      ),
    );
  }

  Widget _buildAddHighlight() {
    return Padding(
      padding: EdgeInsets.only(right: 12),
      child: GestureDetector(
        onTap: _addHighlight,
        child: Column(
          children: [
            Container(
              width: 64,
              height: 64,
              decoration: BoxDecoration(
                shape: BoxShape.circle,
                border: Border.all(color: Colors.grey, width: 1),
              ),
              child: Icon(Icons.add, color: Colors.grey),
            ),
            SizedBox(height: 4),
            Text(
              'New',
              style: TextStyle(fontSize: 12),
              maxLines: 1,
              overflow: TextOverflow.ellipsis,
            ),
          ],
        ),
      ),
    );
  }

  Widget _buildHighlightItem(HighlightStory highlight) {
    return Padding(
      padding: EdgeInsets.only(right: 12),
      child: GestureDetector(
        onTap: () => _viewHighlight(highlight),
        child: Column(
          children: [
            Container(
              width: 64,
              height: 64,
              decoration: BoxDecoration(
                shape: BoxShape.circle,
                border: Border.all(color: Colors.grey, width: 1),
              ),
              child: ClipOval(
                child: Image.network(
                  highlight.cover,
                  fit: BoxFit.cover,
                  errorBuilder: (context, error, stackTrace) {
                    return Container(
                      color: Colors.grey[300],
                      child: Icon(Icons.image, color: Colors.grey),
                    );
                  },
                ),
              ),
            ),
            SizedBox(height: 4),
            Text(
              highlight.title,
              style: TextStyle(fontSize: 12),
              maxLines: 1,
              overflow: TextOverflow.ellipsis,
            ),
          ],
        ),
      ),
    );
  }

  Widget _buildTabBar() {
    return TabBar(
      controller: _tabController,
      labelColor: Colors.black,
      unselectedLabelColor: Colors.grey,
      indicatorColor: Colors.black,
      tabs: [
        Tab(icon: Icon(Icons.grid_on)),
        Tab(icon: Icon(Icons.video_library)),
        Tab(icon: Icon(Icons.person_pin)),
      ],
    );
  }

  Widget _buildPostsGrid() {
    return GridView.builder(
      padding: EdgeInsets.all(2),
      gridDelegate: SliverGridDelegateWithFixedCrossAxisCount(
        crossAxisCount: 3,
        crossAxisSpacing: 2,
        mainAxisSpacing: 2,
      ),
      itemCount: userPosts.length,
      itemBuilder: (context, index) {
        return GestureDetector(
          onTap: () => _viewPost(userPosts[index]),
          child: Image.network(
            userPosts[index].imageUrl,
            fit: BoxFit.cover,
            errorBuilder: (context, error, stackTrace) {
              return Container(
                color: Colors.grey[300],
                child: Icon(Icons.image, color: Colors.grey),
              );
            },
          ),
        );
      },
    );
  }

  Widget _buildReelsGrid() {
    return GridView.builder(
      padding: EdgeInsets.all(2),
      gridDelegate: SliverGridDelegateWithFixedCrossAxisCount(
        crossAxisCount: 3,
        crossAxisSpacing: 2,
        mainAxisSpacing: 2,
        childAspectRatio: 0.75,
      ),
      itemCount: 6, // Sample reels
      itemBuilder: (context, index) {
        return Stack(
          children: [
            Container(
              width: double.infinity,
              height: double.infinity,
              color: Colors.grey[300],
              child: Icon(Icons.play_arrow, size: 30, color: Colors.white),
            ),
            Positioned(
              bottom: 4,
              left: 4,
              child: Row(
                children: [
                  Icon(Icons.play_arrow, color: Colors.white, size: 16),
                  Text(
                    '${(index + 1) * 1000}',
                    style: TextStyle(color: Colors.white, fontSize: 12),
                  ),
                ],
              ),
            ),
          ],
        );
      },
    );
  }

  Widget _buildTaggedGrid() {
    return Center(
      child: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        children: [
          Icon(Icons.person_pin, size: 64, color: Colors.grey),
          SizedBox(height: 16),
          Text(
            'No Photos',
            style: TextStyle(
              fontSize: 24,
              fontWeight: FontWeight.bold,
              color: Colors.grey,
            ),
          ),
          Text(
            'Photos of you will appear here',
            style: TextStyle(color: Colors.grey),
          ),
        ],
      ),
    );
  }

  String _formatCount(int count) {
    if (count < 1000) return count.toString();
    if (count < 1000000) return '${(count / 1000).toStringAsFixed(1)}K';
    return '${(count / 1000000).toStringAsFixed(1)}M';
  }

  void _showOptionsMenu() {
    showModalBottomSheet(
      context: context,
      builder: (context) {
        return Container(
          padding: EdgeInsets.symmetric(vertical: 20),
          child: Column(
            mainAxisSize: MainAxisSize.min,
            children: [
              ListTile(
                leading: Icon(Icons.settings),
                title: Text('Settings'),
                onTap: () {
                  Navigator.pop(context);
                  // Navigate to settings
                },
              ),
              ListTile(
                leading: Icon(Icons.archive),
                title: Text('Archive'),
                onTap: () {
                  Navigator.pop(context);
                  // Navigate to archive
                },
              ),
              ListTile(
                leading: Icon(Icons.history),
                title: Text('Your Activity'),
                onTap: () {
                  Navigator.pop(context);
                  // Navigate to activity
                },
              ),
              ListTile(
                leading: Icon(Icons.qr_code),
                title: Text('QR Code'),
                onTap: () {
                  Navigator.pop(context);
                  // Show QR code
                },
              ),
            ],
          ),
        );
      },
    );
  }

  void _showUserOptionsMenu() {
    showModalBottomSheet(
      context: context,
      builder: (context) {
        return Container(
          padding: EdgeInsets.symmetric(vertical: 20),
          child: Column(
            mainAxisSize: MainAxisSize.min,
            children: [
              ListTile(
                leading: Icon(Icons.report),
                title: Text('Report'),
                onTap: () {
                  Navigator.pop(context);
                  _reportUser();
                },
              ),
              ListTile(
                leading: Icon(Icons.block),
                title: Text('Block'),
                onTap: () {
                  Navigator.pop(context);
                  _blockUser();
                },
              ),
              ListTile(
                leading: Icon(Icons.copy),
                title: Text('Copy Profile URL'),
                onTap: () {
                  Navigator.pop(context);
                  _copyProfileUrl();
                },
              ),
            ],
          ),
        );
      },
    );
  }

  void _editProfile() {
    Navigator.pushNamed(context, '/edit-profile');
  }

  void _shareProfile() {
    ScaffoldMessenger.of(context).showSnackBar(
      SnackBar(content: Text('Profile shared')),
    );
  }

  void _toggleFollow() {
    setState(() {
      isFollowing = !isFollowing;
    });
  }

  void _sendMessage() {
    Navigator.pushNamed(context, '/chat', arguments: widget.username);
  }

  void _openWebsite() {
    ScaffoldMessenger.of(context).showSnackBar(
      SnackBar(content: Text('Opening ${userProfile.website}')),
    );
  }

  void _addHighlight() {
    ScaffoldMessenger.of(context).showSnackBar(
      SnackBar(content: Text('Add highlight functionality')),
    );
  }

  void _viewHighlight(HighlightStory highlight) {
    Navigator.pushNamed(context, '/stories', arguments: highlight);
  }

  void _viewPost(Post post) {
    Navigator.pushNamed(context, '/post-detail', arguments: post);
  }

  void _reportUser() {
    ScaffoldMessenger.of(context).showSnackBar(
      SnackBar(content: Text('User reported')),
    );
  }

  void _blockUser() {
    ScaffoldMessenger.of(context).showSnackBar(
      SnackBar(content: Text('User blocked')),
    );
  }

  void _copyProfileUrl() {
    ScaffoldMessenger.of(context).showSnackBar(
      SnackBar(content: Text('Profile URL copied to clipboard')),
    );
  }

  @override
  void dispose() {
    _tabController.dispose();
    super.dispose();
  }
}

class UserProfile {
  final String username;
  final String displayName;
  final String bio;
  final String avatar;
  final int postsCount;
  final int followersCount;
  final int followingCount;
  final bool isVerified;
  final String website;

  UserProfile({
    required this.username,
    required this.displayName,
    required this.bio,
    required this.avatar,
    required this.postsCount,
    required this.followersCount,
    required this.followingCount,
    required this.isVerified,
    required this.website,
  });
}

class Post {
  final String id;
  final String imageUrl;
  final String caption;
  final int likes;
  final int comments;

  Post({
    required this.id,
    required this.imageUrl,
    required this.caption,
    required this.likes,
    required this.comments,
  });
}

class HighlightStory {
  final String id;
  final String title;
  final String cover;

  HighlightStory({
    required this.id,
    required this.title,
    required this.cover,
  });
}

File: snippets/screens/swiftui/user_profile_screen/3/user_profile_screen.swift
import SwiftUI

struct UserProfileScreen: View {
    let username: String
    @StateObject private var viewModel = UserProfileViewModel()
    @State private var selectedTab = 0

    var body: some View {
        NavigationView {
            ScrollView {
                VStack(spacing: 0) {
                    // Profile header
                    ProfileHeaderView(
                        profile: viewModel.userProfile,
                        isCurrentUser: viewModel.isCurrentUser,
                        isFollowing: viewModel.isFollowing,
                        onEditProfile: { viewModel.editProfile() },
                        onShareProfile: { viewModel.shareProfile() },
                        onToggleFollow: { viewModel.toggleFollow() },
                        onMessage: { viewModel.sendMessage() },
                        onFollowers: { /* Navigate to followers */ },
                        onFollowing: { /* Navigate to following */ }
                    )

                    // Bio section
                    if !viewModel.userProfile.bio.isEmpty {
                        BioSectionView(
                            displayName: viewModel.userProfile.displayName,
                            bio: viewModel.userProfile.bio,
                            website: viewModel.userProfile.website
                        )
                    }

                    // Highlights
                    HighlightsSectionView(
                        highlights: viewModel.highlights,
                        isCurrentUser: viewModel.isCurrentUser,
                        onAddHighlight: { viewModel.addHighlight() },
                        onViewHighlight: { highlight in
                            // Navigate to stories
                        }
                    )

                    // Tab bar
                    TabBarView(selectedTab: $selectedTab)

                    // Content based on selected tab
                    TabContentView(
                        selectedTab: selectedTab,
                        posts: viewModel.posts,
                        onPostTap: { post in
                            // Navigate to post detail
                        }
                    )
                }
            }
            .navigationTitle(viewModel.userProfile.username)
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                if viewModel.isCurrentUser {
                    ToolbarItem(placement: .navigationBarTrailing) {
                        Button(action: { viewModel.showOptionsMenu() }) {
                            Image(systemName: "line.3.horizontal")
                        }
                    }
                } else {
                    ToolbarItem(placement: .navigationBarTrailing) {
                        Menu {
                            Button("Report") { viewModel.reportUser() }
                            Button("Block") { viewModel.blockUser() }
                            Button("Copy Profile URL") { viewModel.copyProfileUrl() }
                        } label: {
                            Image(systemName: "ellipsis")
                        }
                    }
                }
            }
        }
        .onAppear {
            viewModel.loadUserProfile(username: username)
        }
    }
}

struct ProfileHeaderView: View {
    let profile: UserProfile
    let isCurrentUser: Bool
    let isFollowing: Bool
    let onEditProfile: () -> Void
    let onShareProfile: () -> Void
    let onToggleFollow: () -> Void
    let onMessage: () -> Void
    let onFollowers: () -> Void
    let onFollowing: () -> Void

    var body: some View {
        VStack(spacing: 16) {
            HStack {
                // Profile picture
                AsyncImage(url: URL(string: profile.avatar)) { image in
                    image
                        .resizable()
                        .aspectRatio(contentMode: .fill)
                } placeholder: {
                    Circle()
                        .fill(Color.gray.opacity(0.3))
                }
                .frame(width: 80, height: 80)
                .clipShape(Circle())

                Spacer()

                // Stats
                HStack(spacing: 32) {
                    StatView(label: "Posts", count: profile.postsCount)

                    Button(action: onFollowers) {
                        StatView(label: "Followers", count: profile.followersCount)
                    }

                    Button(action: onFollowing) {
                        StatView(label: "Following", count: profile.followingCount)
                    }
                }
            }

            // Action buttons
            if isCurrentUser {
                HStack(spacing: 12) {
                    Button("Edit Profile", action: onEditProfile)
                        .buttonStyle(.bordered)
                        .frame(maxWidth: .infinity)

                    Button("Share Profile", action: onShareProfile)
                        .buttonStyle(.bordered)
                        .frame(maxWidth: .infinity)
                }
            } else {
                HStack(spacing: 12) {
                    Button(isFollowing ? "Following" : "Follow") {
                        onToggleFollow()
                    }
                    .buttonStyle(.borderedProminent)
                    .frame(maxWidth: .infinity)

                    Button("Message", action: onMessage)
                        .buttonStyle(.bordered)
                        .frame(maxWidth: .infinity)
                }
            }
        }
        .padding(.horizontal, 16)
        .padding(.vertical, 12)
    }
}

struct StatView: View {
    let label: String
    let count: Int

    var body: some View {
        VStack(spacing: 2) {
            Text(formatCount(count))
                .font(.headline)
                .fontWeight(.bold)

            Text(label)
                .font(.caption)
                .foregroundColor(.secondary)
        }
    }

    private func formatCount(_ count: Int) -> String {
        if count < 1000 { return "\(count)" }
        if count < 1000000 { return String(format: "%.1fK", Double(count) / 1000) }
        return String(format: "%.1fM", Double(count) / 1000000)
    }
}

struct BioSectionView: View {
    let displayName: String
    let bio: String
    let website: String

    var body: some View {
        VStack(alignment: .leading, spacing: 4) {
            if !displayName.isEmpty {
                Text(displayName)
                    .font(.subheadline)
                    .fontWeight(.semibold)
            }

            Text(bio)
                .font(.subheadline)

            if !website.isEmpty {
                Button(website) {
                    // Open website
                }
                .font(.subheadline)
                .foregroundColor(.blue)
            }
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(.horizontal, 16)
        .padding(.bottom, 12)
    }
}

struct HighlightsSectionView: View {
    let highlights: [HighlightStory]
    let isCurrentUser: Bool
    let onAddHighlight: () -> Void
    let onViewHighlight: (HighlightStory) -> Void

    var body: some View {
        ScrollView(.horizontal, showsIndicators: false) {
            HStack(spacing: 16) {
                if isCurrentUser {
                    Button(action: onAddHighlight) {
                        VStack(spacing: 4) {
                            Circle()
                                .stroke(Color.gray, lineWidth: 1)
                                .frame(width: 64, height: 64)
                                .overlay(
                                    Image(systemName: "plus")
                                        .foregroundColor(.gray)
                                )

                            Text("New")
                                .font(.caption)
                                .foregroundColor(.primary)
                        }
                    }
                }

                ForEach(highlights) { highlight in
                    Button(action: { onViewHighlight(highlight) }) {
                        VStack(spacing: 4) {
                            AsyncImage(url: URL(string: highlight.cover)) { image in
                                image
                                    .resizable()
                                    .aspectRatio(contentMode: .fill)
                            } placeholder: {
                                Circle()
                                    .fill(Color.gray.opacity(0.3))
                            }
                            .frame(width: 64, height: 64)
                            .clipShape(Circle())
                            .overlay(
                                Circle()
                                    .stroke(Color.gray, lineWidth: 1)
                            )

                            Text(highlight.title)
                                .font(.caption)
                                .foregroundColor(.primary)
                                .lineLimit(1)
                        }
                    }
                }
            }
            .padding(.horizontal, 16)
        }
        .padding(.vertical, 8)
    }
}

struct TabBarView: View {
    @Binding var selectedTab: Int

    var body: some View {
        HStack {
            TabButton(
                icon: "grid",
                isSelected: selectedTab == 0,
                action: { selectedTab = 0 }
            )

            TabButton(
                icon: "play.rectangle",
                isSelected: selectedTab == 1,
                action: { selectedTab = 1 }
            )

            TabButton(
                icon: "person.crop.rectangle",
                isSelected: selectedTab == 2,
                action: { selectedTab = 2 }
            )

            Spacer()
        }
        .padding(.horizontal, 16)
        .padding(.vertical, 8)
        .background(Color(UIColor.systemBackground))
    }
}

struct TabButton: View {
    let icon: String
    let isSelected: Bool
    let action: () -> Void

    var body: some View {
        Button(action: action) {
            Image(systemName: icon)
                .font(.title2)
                .foregroundColor(isSelected ? .primary : .secondary)
        }
        .frame(maxWidth: .infinity)
    }
}

struct TabContentView: View {
    let selectedTab: Int
    let posts: [UserPost]
    let onPostTap: (UserPost) -> Void

    var body: some View {
        Group {
            switch selectedTab {
            case 0:
                PostsGridView(posts: posts, onPostTap: onPostTap)
            case 1:
                ReelsGridView()
            case 2:
                TaggedGridView()
            default:
                EmptyView()
            }
        }
    }
}

struct PostsGridView: View {
    let posts: [UserPost]
    let onPostTap: (UserPost) -> Void

    let columns = Array(repeating: GridItem(.flexible(), spacing: 2), count: 3)

    var body: some View {
        LazyVGrid(columns: columns, spacing: 2) {
            ForEach(posts) { post in
                Button(action: { onPostTap(post) }) {
                    AsyncImage(url: URL(string: post.imageUrl)) { image in
                        image
                            .resizable()
                            .aspectRatio(1, contentMode: .fill)
                    } placeholder: {
                        Rectangle()
                            .fill(Color.gray.opacity(0.3))
                            .overlay(
                                Image(systemName: "photo")
                                    .foregroundColor(.gray)
                            )
                    }
                    .clipped()
                }
            }
        }
        .padding(.horizontal, 2)
    }
}

struct ReelsGridView: View {
    let columns = Array(repeating: GridItem(.flexible(), spacing: 2), count: 3)

    var body: some View {
        LazyVGrid(columns: columns, spacing: 2) {
            ForEach(0..<6) { index in
                Rectangle()
                    .fill(Color.gray.opacity(0.3))
                    .aspectRatio(0.75, contentMode: .fill)
                    .overlay(
                        VStack {
                            Spacer()
                            HStack {
                                Image(systemName: "play.fill")
                                    .foregroundColor(.white)
                                    .font(.caption)
                                Text("\((index + 1) * 1000)")
                                    .foregroundColor(.white)
                                    .font(.caption)
                                Spacer()
                            }
                            .padding(4)
                        }
                    )
            }
        }
        .padding(.horizontal, 2)
    }
}

struct TaggedGridView: View {
    var body: some View {
        VStack(spacing: 16) {
            Image(systemName: "person.crop.rectangle")
                .font(.system(size: 64))
                .foregroundColor(.gray)

            VStack(spacing: 4) {
                Text("No Photos")
                    .font(.title2)
                    .fontWeight(.bold)
                    .foregroundColor(.gray)

                Text("Photos of you will appear here")
                    .font(.subheadline)
                    .foregroundColor(.secondary)
            }
        }
        .frame(maxWidth: .infinity, maxHeight: 300)
    }
}

@MainActor
class UserProfileViewModel: ObservableObject {
    @Published var userProfile = UserProfile(
        username: "john_doe",
        displayName: "John Doe",
        bio: "Photographer 📸 | Travel enthusiast ✈️ | Coffee lover ☕️\nCapturing moments around the world",
        avatar: "https://example.com/avatar1.jpg",
        postsCount: 127,
        followersCount: 1524,
        followingCount: 892,
        isVerified: true,
        website: "www.johndoe.photography"
    )

    @Published var posts: [UserPost] = []
    @Published var highlights: [HighlightStory] = [
        HighlightStory(id: "1", title: "Travel", cover: "https://example.com/highlight1.jpg"),
        HighlightStory(id: "2", title: "Food", cover: "https://example.com/highlight2.jpg"),
        HighlightStory(id: "3", title: "Work", cover: "https://example.com/highlight3.jpg")
    ]

    @Published var isFollowing = false
    @Published var isCurrentUser = false

    func loadUserProfile(username: String) {
        isCurrentUser = username == "current_user"

        // Load sample posts
        posts = [
            UserPost(id: "1", imageUrl: "https://example.com/post1.jpg", caption: "Beautiful sunset", likes: 156, comments: 23),
            UserPost(id: "2", imageUrl: "https://example.com/post2.jpg", caption: "Morning coffee", likes: 89, comments: 12),
            UserPost(id: "3", imageUrl: "https://example.com/post3.jpg", caption: "City lights", likes: 234, comments: 45),
            UserPost(id: "4", imageUrl: "https://example.com/post4.jpg", caption: "Mountain view", likes: 312, comments: 67),
            UserPost(id: "5", imageUrl: "https://example.com/post5.jpg", caption: "Beach day", likes: 198, comments: 34),
            UserPost(id: "6", imageUrl: "https://example.com/post6.jpg", caption: "Street art", likes: 156, comments: 28)
        ]
    }

    func editProfile() {
        // Navigate to edit profile
    }

    func shareProfile() {
        // Share profile functionality
    }

    func toggleFollow() {
        isFollowing.toggle()
    }

    func sendMessage() {
        // Navigate to chat
    }

    func showOptionsMenu() {
        // Show options menu
    }

    func reportUser() {
        // Report user functionality
    }

    func blockUser() {
        // Block user functionality
    }

    func copyProfileUrl() {
        // Copy profile URL
    }

    func addHighlight() {
        // Add new highlight
    }
}

struct UserProfile {
    let username: String
    let displayName: String
    let bio: String
    let avatar: String
    let postsCount: Int
    let followersCount: Int
    let followingCount: Int
    let isVerified: Bool
    let website: String
}

struct UserPost: Identifiable {
    let id: String
    let imageUrl: String
    let caption: String
    let likes: Int
    let comments: Int
}

struct HighlightStory: Identifiable {
    let id: String
    let title: String
    let cover: String
}

#Preview {
    UserProfileScreen(username: "john_doe")
}